histogram/src/standard.rs

```diff
@@ -0,0 +1,253 @@
+
+use crate::{Bucket, BuildError, Config, Error};
+
+/// A simple histogram that can be used to track the distribution of occurrences
+/// of quantized u64 values.
+///
+/// Internally it uses 64bit counters to store the counts for each bucket.
+pub struct Histogram {
+    pub(crate) config: Config,
+    pub(crate) total_count: u128,
+    pub(crate) buckets: Box<[u64]>,
+}
+
+impl Histogram {
+    /// Construct a new `Histogram` from the provided parameters.
+    /// * `a` sets bin width in the linear portion, the bin width is `2^a`
+    /// * `b` sets the number of divisions in the logarithmic portion to `2^b`.
+    /// * `n` sets the max value as `2^n`. Note: when `n` is 64, the max value
+    ///   is `u64::MAX`
+    ///
+    /// # Constraints
+    /// * `n` must be less than or equal to 64
+    /// * `n` must be greater than `a + b`
+    pub fn new(a: u8, b: u8, n: u8) -> Result<Self, BuildError> {
+        let config = Config::new(a, b, n)?;
+
+        Ok(Self::from_config(config))
+    }
+
+    /// Increment the counter for the bucket corresponding to the provided value
+    /// by one.
+    pub fn increment(&mut self, value: u64) -> Result<(), Error> {
+        self.add(value, 1)
+    }
+
+    /// Add some count to the counter for the bucket corresponding to the
+    /// provided value
+    pub fn add(&mut self, value: u64, count: u64) -> Result<(), Error> {
+        let index = self.config.value_to_index(value)?;
+        self.buckets[index] = self.buckets[index].wrapping_add(count);
+        self.total_count = self.total_count.wrapping_add(count.into());
+        Ok(())
+    }
+
+    /// Creates a new `Histogram` from the `Config`.
+    pub(crate) fn from_config(config: Config) -> Self {
+        let buckets: Box<[u64]> = vec![0; config.total_bins()].into();
+
+        Self {
+            config,
+            total_count: 0,
+            buckets,
+        }
+    }
+
+    /// Get a reference to the raw counters.
+    pub fn as_slice(&self) -> &[u64] {
+        &self.buckets
+    }
+
+    /// Get a mutable reference to the raw counters.
+    pub fn as_mut_slice(&mut self) -> &mut [u64] {
+        &mut self.buckets
+    }
+
+    /// Return a collection of percentiles from this histogram.
+    ///
+    /// Each percentile should be in the inclusive range `0.0..=100.0`. For
+    /// example, the 50th percentile (median) can be found using `50.0`.
+    ///
+    /// The results will be sorted by the percentile.
+    pub(crate) fn percentiles(&self, percentiles: &[f64]) -> Result<Vec<(f64, Bucket)>, Error> {
+        // if the histogram is empty, then we should return an error
+        if self.total_count == 0_u128 {
+            return Err(Error::Empty);
+        }
+
+        // sort the requested percentiles so we can find them in a single pass
+        let mut percentiles = percentiles.to_vec();
+        percentiles.sort_by(|a, b| a.partial_cmp(b).unwrap());
+
+        // validate all the percentiles
+        for percentile in &percentiles {
```

Comment:
This doesn't need to loop. We can just check the last percentile.

Reply:
And the first presumably, since it is a floating point?

Reply:
Yes. And pushed the change for this.